File: internal/infra/model/vm.go
// Package model consolidates ALL model types from across the Go codebase into
// a single shared package. No domain imports anything outside the model package.
package model

// ── Status (VM lifecycle) ──

// Status is the VM lifecycle status type, matching Python's VMStatus(StrEnum).
type Status string

const (
	StatusStarting Status = "starting"
	StatusRunning  Status = "running"
	StatusPaused   Status = "paused"
	StatusStopping Status = "stopping"
	StatusStopped  Status = "stopped"
	StatusCrashed  Status = "crashed"
	StatusError    Status = "error"
)

// ── VM ──

// VM matches Python's VMInstanceItem dataclass exactly.
type VM struct {
	ID            string `json:"id" db:"id"`
	Name          string `json:"name" db:"name"`
	Status        Status `json:"status" db:"status"`
	PID           int    `json:"pid" db:"pid"`
	IPv4          string `json:"ipv4" db:"ipv4"`
	MAC           string `json:"mac" db:"mac"`
	NetworkID     string `json:"network_id" db:"network_id"`
	TapDevice     string `json:"tap_device" db:"tap_device"`
	ImageID       string `json:"image_id" db:"image_id"`
	KernelID      string `json:"kernel_id" db:"kernel_id"`
	BinaryID      string `json:"binary_id" db:"binary_id"`
	APISocketPath string `json:"api_socket_path" db:"api_socket_path"`
	ConfigPath    string `json:"config_path" db:"config_path"`
	CloudInitMode string `json:"cloud_init_mode" db:"cloud_init_mode"`
	VCPUCount     int    `json:"vcpu_count" db:"vcpu_count"`
	MemSizeMiB    int    `json:"mem_size_mib" db:"mem_size_mib"`
	DiskSizeMiB   int    `json:"disk_size_mib" db:"disk_size_mib"`
	RootfsPath    string `json:"rootfs_path" db:"rootfs_path"`
	RootfsSuffix  string `json:"rootfs_suffix" db:"rootfs_suffix"`
	PCIEnabled    bool   `json:"pci_enabled" db:"pci_enabled"`
	NestedVirt    bool   `json:"nested_virt" db:"nested_virt"`
	EnableLogging bool   `json:"enable_logging" db:"enable_logging"`
	EnableMetrics bool   `json:"enable_metrics" db:"enable_metrics"`
	EnableConsole bool   `json:"enable_console" db:"enable_console"`
	CreatedAt     string `json:"created_at" db:"created_at"`
	UpdatedAt     string `json:"updated_at" db:"updated_at"`

	// Optional fields
	RelaySocketPath  *string `json:"relay_socket_path,omitempty" db:"relay_socket_path"`
	ProcessStartTime *int64  `json:"process_start_time,omitempty" db:"process_start_time"`
	NocloudNetPort   *int    `json:"nocloud_net_port,omitempty" db:"nocloud_net_port"`
	NocloudNetPID    *int    `json:"nocloud_net_pid,omitempty" db:"nocloud_net_pid"`
	RelayPID         *int    `json:"relay_pid,omitempty" db:"relay_pid"`
	ExitCode         *int    `json:"exit_code,omitempty" db:"exit_code"`
	LogPath          *string `json:"log_path,omitempty" db:"log_path"`
	SerialOutputPath *string `json:"serial_output_path,omitempty" db:"serial_output_path"`
	LSMFlags         *string `json:"lsm_flags,omitempty" db:"lsm_flags"`
	BootArgs         *string `json:"boot_args,omitempty" db:"boot_args"`

	// JSON-serialized in DB fields
	SSHKeys   []string   `json:"ssh_keys"`
	SSHUser   *string    `json:"ssh_user,omitempty" db:"ssh_user"`
	VolumeIDs []string   `json:"volume_ids,omitempty"`
	CPUConfig *CpuConfig `json:"cpu_config,omitempty"`

	// Resolved relations (typed as concrete model types from this package)
	Kernel  *KernelItem   `json:"kernel,omitempty"`
	Image   *ImageItem    `json:"image,omitempty"`
	Binary  *BinaryItem   `json:"binary,omitempty"`
	Network *Network      `json:"network,omitempty"`
	Volumes []*VolumeItem `json:"volumes,omitempty"`
}

// ── ConsoleInfo ──

// ConsoleInfo matches Python's ConsoleInfo dataclass.
type ConsoleInfo struct {
	SocketPath string `json:"socket_path"`
	VMName     string `json:"vm_name"`
}

// ── ConsoleState ──

// ConsoleState matches Python's ConsoleState dataclass.
type ConsoleState struct {
	Running    bool    `json:"running"`
	PID        *int    `json:"pid,omitempty"`
	SocketPath *string `json:"socket_path,omitempty"`
}

// ── VMInspectInfo ──

// VMInspectInfo matches Python's VMInspectInfo dataclass.
type VMInspectInfo struct {
	ID            string             `json:"id"`
	Name          string             `json:"name"`
	Status        string             `json:"status"`
	CreatedAt     *string            `json:"created_at,omitempty"`
	PID           *int               `json:"pid,omitempty"`
	IP            *string            `json:"ip,omitempty"`
	MAC           *string            `json:"mac,omitempty"`
	NetworkName   *string            `json:"network_name,omitempty"`
	TapDevice     *string            `json:"tap_device,omitempty"`
	CloudInitMode string             `json:"cloud_init_mode"`
	ImageID       *string            `json:"image_id,omitempty"`
	ImageName     *string            `json:"image_name,omitempty"`
	KernelID      *string            `json:"kernel_id,omitempty"`
	KernelName    *string            `json:"kernel_name,omitempty"`
	Paths         map[string]*string `json:"paths"`
	Features      map[string]bool    `json:"features"`
	NocloudNet    map[string]any     `json:"nocloud_net,omitempty"` // Flexible: shape varies by cloud-init provider (NoCloud, OpenStack, etc.)
	Console       map[string]any     `json:"console,omitempty"`     // Flexible: console state fields differ by relay implementation
}

File: legacy/python/src/mvmctl/models/vm.py
"""VM data models."""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from enum import StrEnum, auto
from pathlib import Path
from typing import Any

from mvmctl.models.binary import BinaryItem
from mvmctl.models.firecracker import CpuConfig
from mvmctl.models.image import ImageItem
from mvmctl.models.kernel import KernelItem
from mvmctl.models.network import NetworkItem
from mvmctl.models.volume import VolumeItem
from mvmctl.utils.common import CommonUtils


class VMStatus(StrEnum):
    """VM lifecycle states."""

    STARTING = auto()
    RUNNING = auto()
    PAUSED = auto()
    STOPPING = auto()
    STOPPED = auto()
    CRASHED = auto()
    ERROR = auto()


@dataclass
class VMInstanceItem:
    """VM instance record — maps to vm_instances table."""

    id: str
    name: str
    status: str
    pid: int
    ipv4: str
    mac: str
    network_id: str
    tap_device: str
    image_id: str
    kernel_id: str
    binary_id: str
    api_socket_path: str
    config_path: str
    cloud_init_mode: str
    vcpu_count: int
    mem_size_mib: int
    disk_size_mib: int
    rootfs_path: str
    rootfs_suffix: str
    pci_enabled: bool
    nested_virt: bool
    enable_logging: bool
    enable_metrics: bool
    enable_console: bool
    created_at: str
    updated_at: str

    relay_socket_path: str | None = None
    process_start_time: int | None = None
    nocloud_net_port: int | None = None
    nocloud_net_pid: int | None = None
    relay_pid: int | None = None
    exit_code: int | None = None
    log_path: str | None = None
    serial_output_path: str | None = None
    lsm_flags: str | None = None
    boot_args: str | None = None
    ssh_keys: list[str] = field(
        default_factory=list
    )  # SSH key fingerprints stored in VM
    ssh_user: str | None = None  # SSH user for this VM
    volume_ids: list[str] | None = None  # Attached volume IDs
    cpu_config: CpuConfig | None = None  # JSON: merged CPU template config

    def __post_init__(self) -> None:
        """Deserialize ssh_keys, volume_ids, and cpu_config from JSON strings when loading from DB."""
        if isinstance(self.ssh_keys, str):
            self.ssh_keys = json.loads(self.ssh_keys)
        if isinstance(self.volume_ids, str):
            self.volume_ids = json.loads(self.volume_ids)
        if isinstance(self.cpu_config, str):
            self.cpu_config = json.loads(self.cpu_config)
        CommonUtils.coerce_bool_fields(
            self,
            {
                "pci_enabled",
                "nested_virt",
                "enable_logging",
                "enable_metrics",
                "enable_console",
            },
        )

    # Resolved relations
    kernel: KernelItem | None = None
    image: ImageItem | None = None
    binary: BinaryItem | None = None
    network: NetworkItem | None = None
    volumes: list[VolumeItem] = field(default_factory=list)

    @property
    def vm_dir(self) -> Path:
        """Absolute VM directory path."""
        from mvmctl.utils.common import CacheUtils

        return CacheUtils.get_vm_dir(self.id)


@dataclass
class ConsoleInfo:
    """Information about a VM console socket."""

    socket_path: Path
    vm_name: str


@dataclass
class ConsoleState:
    """Current state of the console relay process."""

    running: bool
    pid: int | None
    socket_path: str | None


@dataclass
class VMInspectInfo:
    """Complete inspection output for a VM."""

    id: str
    name: str
    status: str
    created_at: str | None
    pid: int | None
    ip: str | None
    mac: str | None
    network_name: str | None
    tap_device: str | None
    cloud_init_mode: str
    image_id: str | None
    image_name: str | None
    kernel_id: str | None
    kernel_name: str | None
    paths: dict[str, str | None]  # vm_dir, rootfs, config
    features: dict[str, bool]  # api_socket, console, nocloud_net
    nocloud_net: dict[str, Any] | None = None
    console: dict[str, Any] | None = None
